crates/amaru-ledger/src/rules/transaction/inputs.rs

```diff
@@ -63,15 +64,31 @@ where
                 intersection.push(reference_input.clone());
             }
 
-            if let Some(output) = context.lookup(reference_input) {
-                if let Some(BorrowedDatumOption::Hash(hash)) = output.datum() {
-                    context.allow_supplemental_datum(*hash);
-                }
-            } else {
-                return Err(InvalidInputs::UnknownInput(reference_input.clone()));
+            // We have to drop the output before mutating context, otherwise we have a mutable borrow and an immutable borrow at the same time
+            let (datum_hash, script_ref) = {
+                let output = context
+                    .lookup(reference_input)
+                    .ok_or_else(|| InvalidInputs::UnknownInput(reference_input.clone()))?;
+
+                let datum_hash = match output.datum() {
+                    Some(BorrowedDatumOption::Hash(hash)) => Some(*hash),
+                    _ => None,
+                };
+
+                let script_ref = output.has_script_ref().cloned();
```

Comment:
We are cloning the script here, which is *really* not ideal. Scripts and datums are the worst things we can clone during the validation process.

That being said, the context can't have a reference to the script right now, as the context has a lifetime that is longer than the script. We may potentially be able to adjust that, since the lifetime of the scripts should come from the lifetime of the block (??), but that is entering a lifetime hell I'm not prepared for right now.

Really open to ideas here, I can't think of a solution.

Reply:
I'll leave a FIXME so that this concern is documented in the code.